torch/csrc/jit/passes/break_continue_transform.cpp

```diff
@@ -0,0 +1,326 @@
+  }
+
+  // Recurses on the if node and returns its return status
+  // If status != WONT, sets the block_return_val and sentinel val
+  // of its parent block before exit
+  LoopStatus handleIf(Node* node) {
+    auto true_block = node->blocks().at(0);
+    auto false_block = node->blocks().at(1);
+
+    auto true_status = handleTransforms(true_block);
+    auto false_status = handleTransforms(false_block);
+
+    if (true_status == WONT && false_status == WONT) {
+      return WONT;
+    } else if (true_status == WILL && false_status == WILL) {
+      return WILL;
+    } else {
+      return MIGHT;
+    }
+  }
+
+  LoopStatus guardBlockNodes(
+      Block* block,
+      generic_graph_node_list_iterator<Node>& iter) {
+    // if an if node might hit a break or continue statement,
+    // we guard all subsequent nodes in the block, and only execute them
+    // if we did break / did continue is false.
+
+    auto new_if = graph_->create(prim::If, 0)->insertBefore(*iter);
+    auto sentinel =
+        graph_->createLoad(getVarname(), BoolType::get())->insertBefore(new_if);
+    new_if->addInput(sentinel->output());
+
+    auto hit_control_flow_block = new_if->addBlock();
+    auto guard_block = new_if->addBlock();
+
+    while (iter != block->nodes().end()) {
+      auto node = *iter++;
+      node->moveBefore(guard_block->return_node());
+    }
+
+    {
+      WithInsertPoint insert(hit_control_flow_block);
+      // NB: insert var scape before transform kind so it is not removed
+      // In a graph like:
+      // for i in range(3):
+      //     if cond == 2:
+      //         if cond == 2:
+      //             m = 2
+      //             break
+      //         k = 1
+      //     else:
+      //         k = 2
+      //     m += k
+      // We transform the inner cond == 2 block to look like:
+      // if cond == 2:
+      //     m = 2
+      //     $did_break = True
+      // else:
+      //     $did_break = False
+      // if $did_break...
+      //    prim::NewVarEscape
+      // else:
+      //    k = 1
+      // For these new if nodes that guard ops after a continue/break may have
+      // occurred, the new variables that are defined need to escape scope.
+      // Otherwise, in the example above, we would error in the m += k call.
+      graph_->insertNode(graph_->create(prim::NewVarEscape, 0));
+      graph_->insertNode(graph_->create(transformKind(), 0));
+    }
+    return handleIf(new_if);
+  }
+
+  void deleteAfterBreakNodes(Block* block, graph_node_list_iterator& iter) {
+    if (iter == block->nodes().end()) {
+      return;
+    }
+    // need to destroy in reverse order so nodes have no uses when destroyed
+    for (auto it = block->nodes().reverse().begin(); it != iter;) {
+      if (*it == block->return_node()) {
+        it++;
+      } else {
+        it.destroyCurrent();
+      }
+    }
+    iter->destroy();
+  }
+
+  void inlineLoopConditionIntoLoopBody(Node* n) {
```

Comment:
Why is this happening in this pass? This is not composable because now to lower loops we have to do the break_continue transform. That is surprising.